Pause MCP before draining/rebooting node

pliurh opened the pull request:
The config daemon instance on each node need to compete for the 'Draining' lock before
it can draining/rebooting a node.

After a config daemon instance gets the lock, it checks its MachineConfigPool conditions.
If the MCP is in ready state, it will pause the MCP then process, otherwise, it wait for
the MCP getting ready. The MCP will be resumed after the config daemon and release the
lock after the config dameon finishes its work.

zshi-redhat commented on pkg/daemon/daemon.go:
```diff
@@ -750,18 +764,84 @@ func (dn *Daemon) annotateNode(node, value string) error {
+			if newMcp.GetName() != dn.mcpName {
+				return
+			}
+			if mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolDegraded) &&
```
When node gets rebooted by sriov config daemon, will the `drainNode` be called again (in which case MCP is degraded due to pause:true)?

pliurh replied:
Pausing an MCP will not make it 'Degraded'.

zshi-redhat commented on pkg/daemon/daemon.go:
```diff
@@ -750,18 +764,84 @@ func (dn *Daemon) annotateNode(node, value string) error {
+			if mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolDegraded) &&
+				mcfgv1.IsMachineConfigPoolConditionTrue(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdated) &&
+				mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdating) {
+				glog.Infof("drainNode(): MCP %s is ready", newMcp.GetName())
```
Check `readyMachineCount` == `machineCount` as well ?

pliurh replied:
I suppose the 'Updated' condition of MCP is sufficient.

zshi-redhat commented on pkg/daemon/daemon.go:
```diff
@@ -750,18 +764,84 @@ func (dn *Daemon) annotateNode(node, value string) error {
+			if newMcp.GetName() != dn.mcpName {
+				return
+			}
+			if mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolDegraded) &&
```
@yuqi-zhang @sinnykumari, do you think the checks here can 100% guarantee there is no race condition between MCO and SNO trying to start an update simultaneously?

pliurh replied:
> By checking poolDegraded, it makes sure pool is not paused by user, right?

No, this is to check MCP is not in 'Degraded' state.

yuqi-zhang replied:
Pausing and degraded are separate concepts. I think the SRIOV operator shouldn't do anything if a pool is degraded (which means the pool is in an error state), as you have here. I think the checks (not degraded, not updating, updated) should be good.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -580,8 +590,17 @@ func (dn *Daemon) completeDrain() error {
 	}
 
+	if utils.ClusterType == utils.ClusterTypeOpenshift {
+		glog.Infof("completeDrain(): resume MCP %s", dn.mcpName)
```
Making sure I understand this properly, the completeDrain() function gets called after the reboot correct? To uncordon the node?

pliurh replied:
Yes. It will be executed after reboot or after draining without a reboot.

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +766,93 @@ func (dn *Daemon) annotateNode(node, value string) error {
+		<-ctx.Done()
+
+		glog.Infof("drainNode(): pause MCP %s", dn.mcpName)
+		pausePatch := []byte("{\"spec\":{\"paused\":true}}")
```
if we want to ensure maximum safety and elimate races, after the pause goes through here, we should add another check like above to see if the pool is progressing, since there is a tiny chance that the MCP started some kind of update in between.

If we check here and the pool is still good, then we can progress with drain. If the pool somehow is updating now, we should consider unpausing and then waiting in a loop for the MCP to go back to not updating.

zshi-redhat replied:
Do we also want to check that pause is not already set by user?

pliurh replied:
I updated the patch, PTAL.

yuqi-zhang reviewed (commented):
Overall this approach I think is more safe with the MCO and race conditions. Some more specific comments below

sinnykumari commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +766,116 @@ func (dn *Daemon) annotateNode(node, value string) error {
+			}
+			if paused {
+				glog.Infof("drainNode(): MCP is processing, resume MCP %s", dn.mcpName)
+				pausePatch := []byte("{\"spec\":{\"paused\":false}}")
```
For my understanding, can you please explain the condition in which we are unpausing the pool here.

pliurh replied:
The MCP informer will not be stoped right after we pause the MCP. So the mcpEventHandler will be invoked once more as we updated the MCD. In this cycle, we check if the MCP is processing, if yes and the MCP was paused during the last cycle (indicated by the variable 'paused' here). We resume the MCP and wait until it finishes.

pliurh replied:
This is to address https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/93#discussion_r599989656

sinnykumari replied:
got it, makes sense.

sinnykumari reviewed (commented):
Thanks for implementing it in this way! overall this approach looks good.

SchSeba commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +766,116 @@ func (dn *Daemon) annotateNode(node, value string) error {
+					cancel()
+					return
+				}
+				if newMcp.Spec.Paused {
```
will this be the case if the sriov operator drains the node and then reboot it to change for example the mlx max number of VFs?

pliurh replied:
After reboot, the config daemon will not drain the node, so we will not hit this logic.

SchSeba replied:
Hi @pliurh did you try this on an intel nic?

I am waiting to get an env but I think in general for intel

1. add the iommu in the kargs
2. drainNode code https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/daemon/daemon.go#L535
3. reboot https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/daemon/daemon.go#L551
4. when the node is up again the default plugin will request to drain again https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/plugins/generic/generic_plugin.go#L86
5. to configure the numOfVfs https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/f552cd8eb8990787a24a0504f633405adfd2af76/pkg/plugins/generic/generic_plugin.go#L193

it will stuck because the mcp is already pause do I miss something?

pliurh replied:
In step5, It will not require to drain a node, as the `ifaceStatus.NumVfs` is always 0 when a node boots up.

However, there is still a case that can trigger the problem, when a non-default MTU is specified. I'll fix that.

SchSeba reviewed (commented):
looks good I just have small comments thanks!

pliurh commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +766,116 @@ func (dn *Daemon) annotateNode(node, value string) error {
 	return nil
 }
 
+func (dn *Daemon) getNodeMachinePool() error {
```
hmm, I think the current approach is ok. @sinnykumari @yuqi-zhang?

sinnykumari replied:
yes, current approach is fine to determine corresponding pool for the node.

pliurh commented on deploy/clusterrole.yaml:
```diff
@@ -48,3 +48,6 @@ rules:
 - apiGroups: [""]
   resources: ["pods/eviction"]
   verbs: ["create"]
+- apiGroups: ["machineconfiguration.openshift.io"]
```
As that is our downstream bundle, I will update it in the openshift repo only.

SchSeba commented:
/lgtm

zshi-redhat commented on pkg/daemon/daemon.go:
```diff
@@ -580,8 +590,17 @@ func (dn *Daemon) completeDrain() error {
+	if utils.ClusterType == utils.ClusterTypeOpenshift {
+		glog.Infof("completeDrain(): resume MCP %s", dn.mcpName)
+		pausePatch := []byte("{\"spec\":{\"paused\":false}}")
+		if _, err := dn.mcClient.MachineconfigurationV1().MachineConfigPools().Patch(context.Background(), dn.mcpName, types.MergePatchType, pausePatch, metav1.PatchOptions{}); err != nil {
```
After rebooting the node, dn.mcpName is empty which results in failure updating the api.
dn.mcpName is assigned in drainNode function, but drainNode is not always called after rebooting.

pliurh replied:
The `dn.mcpName` is assigned in `getNodeMachinePool`, which will be invoked in `nodeStateSyncHandler`. So it will always be assigned before we consume it.

zshi-redhat replied:
I'm testing in a baremetal environment, what I observed was:
```
E0401 01:01:52.621664    5940 daemon.go:592] completeDrain(): failed to resume MCP : resource name may not be empty
E0401 01:01:52.621674    5940 daemon.go:564] nodeStateSyncHandler(): failed to complete draining: resource name may not be empty
```
The mcp never comes back.

zshi-redhat replied:
ignore this, I might use wrong build.

pliurh commented:
@zshi-redhat @SchSeba I just updated the patch. A new annotation value `Draining_MCP_Paused` was introduced to indicate whether the MCP is paused by the config daemon. Also, I changed the `needDrainNode`, so that changing MTU will not need a drain if `ifaceStatus.NumVfs` is 0.

zshi-redhat commented:
/lgtm

SchSeba commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +767,127 @@ func (dn *Daemon) annotateNode(node, value string) error {
+				pausePatch := []byte("{\"spec\":{\"paused\":true}}")
+				_, err = dn.mcClient.MachineconfigurationV1().MachineConfigPools().Patch(context.Background(), dn.mcpName, types.MergePatchType, pausePatch, metav1.PatchOptions{})
+				if err != nil {
+					glog.V(2).Infof("drainNode(): fail to pause MCP %s: %v", dn.mcpName, err)
```
nit: `fail` -> `Failed`

pliurh replied:
fixed.

SchSeba commented on pkg/daemon/daemon.go:
```diff
@@ -747,21 +767,127 @@ func (dn *Daemon) annotateNode(node, value string) error {
-	if err != nil {
-		glog.Errorf("drainNode(): Failed to annotate node: %v", err)
-		return err
+		mcpInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```
nit: can we add a check if this is the MCP that the node is part of even before we send it to the `mcpEventHandler` function?

pliurh replied:
Hmm, I don't know how we can do that. Normally, the informer watches all the objects. I think it's sufficient that we check it in the `mcpEventHandler`.

SchSeba reviewed (commented):
just two small nits and I think the PR is ready
